### ISSUE TITLE
Rendering visualizer to video ability

### ISSUE THREAD

**Issue by iGerman00** [caseif/js.nation#16]

**Comment by AlexVestin** [comment]
I can help with this if there's interest

**Comment by iGerman00** [comment]
There is, there still is.

сб, 19 янв. 2019 г. в 18:01, AlexVestin <notifications@github.com>:

**Comment by AlexVestin** [comment]
@iGerman00 Cool. :) Well we probably need input from one of the maintainers if they want this in their repo, or if we should make a fork

**Comment by iGerman00** [comment]
That would be so awesome, an actual alternative to rendering a visualizer
in after effects, which takes ages, while this will take like, I don’t
know, duration of the song +- a minute. And you can freely customize it to
your liking too, for example I got rid of particles and changed the logo to
another one, and customized the way the waveform looks like a bit (even
though I don’t code, I still managed to)

сб, 19 янв. 2019 г. в 20:36, AlexVestin <notifications@github.com>:

**Comment by caseif** [comment]
I'm not too keen to the idea of merging such a feature into master, as it somewhat undermines the emphasis on realtime rendering. Of course, you're still welcome to fork the project under the terms of the GPL3 (all modifications must remain open-source), and I'd be very interested to see such an implementation in action.

Closing this for the time being.

**Comment by caseif** [comment]
@itsIncept, I'd also be interested in your take on this - I realize I may be closing this a bit preemptively.

**Comment by itsIncept** [comment]
I'd have to agree with @caseif and as an After Effects guy myself, It really isn't a substitute for the accuracy and customizability of an NLE. However if any progress is made in making the base code better we'd definitely look into that. Even with our best efforts it is a little spaghetti in places aha.

**Comment by AlexVestin** [comment]
> It really isn't a substitute for the accuracy and customizability of an NLE

Yeah, I agree. But, I think smaller template-like animations like these work great on the web. 
 
I hacked together a [demo](https://github.com/AlexVestin/js.nation) using the MediaRecorder API, which didn't seem to be able to keep up with the encoding. [Here](https://www.youtube.com/watch?v=ge75IBVbYOI&feature=youtu.be) is an example of an export at 720p, and the quality is pretty bad. 

It could be done with an ffmpeg wasm port (like [mine](https://github.com/AlexVestin/WasmVideoEncoder)), but then there would need to be changes to the structure of the project, since it's too slow for real-time exporting, so would you guys mind me having a stab at porting some of the code to [a template music exporting project](https://github.com/AlexVestin/mvid2)?

**Comment by caseif** [comment]
@AlexVestin As long as it's released under a free license, go for it!

**Comment by AlexVestin** [comment]
@caseif  @itsIncept  I have an early version of the project [here](http://mvid2.org.s3-website.eu-central-1.amazonaws.com/editor?template=Monstercat) if you want to check it out :). I added this project and the vis.js project to it, let me know if you want me to change any of the attribution for them!

**Comment by itsIncept** [comment]
That's pretty cool, can see there's quite a lot of visual things to work out but video downloading worked fine for me.

**Comment by caseif** [comment]
Looks really cool, I like that the visualizer is configurable!

As a note, @itsIncept contributed to vis.js a great deal, so he should be included in the attribution notice as well.

Regarding attribution: technically speaking, output of a GPL-licensed program isn't considered derivative work. This gets a little hairy since the js.nation emblem is copied to the output in this case, which may make it eligible for attribution requirement. Further complicating things are the background images grabbed from Reddit which may or may not free, and the Monstercat emblem in vis.js which is most certainly not free. I'll have to look into this a bit more and get back to you on what exactly is required.

**Comment by AlexVestin** [comment]
I'll add him to the attrib when I get back! The background images are pulled from pexel, so they're free, also the monstercat logo isn't used. 

I can make it so that users load their own logo for the emblem at jsnation. I have a dialogue in place to ensure attribution to the creators from users when exporting items containing licenses that require it, but I will work to add more legally correct instructions before a live push.

**Comment by itsIncept** [comment]
![noticed the emblem ratio looks a little odd on my end](https://user-images.githubusercontent.com/12763419/53665523-c8651a80-3cbf-11e9-829a-6b42cb56b8db.png)

You should probably look into using Montserrat for the MC vis font too, its what they use atm. I'm sure you know the proportions are a bit messed up too, I spent heaps of time on it, also have to take into consideration long titles, double bracketing etc, from what I remember of my personal branch it became a bit of a mess to impliment lol.

**Comment by AlexVestin** [comment]
Yeah I broke the aspect ratios earlier today, should be fixed again, I also imported the Montserrat font for the MC vis. Yeah, the text and logo for the MC vis don't look very good yet. They were done (pretty hastely) today. I plan on adding the boxshadow to the logo-image, and sizing the text better. There's some difficulty porting this since it has to be drawn to a canvas, which leaves out the css implementations. But I also think since the text configs will be able to be edited by users (like placement and size) there can be some wiggle-room on the autoplacement.
